pmd-core/src/main/java/net/sourceforge/pmd/cpd/token/AntlrTokenFilter.java

```diff
@@ -0,0 +1,45 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.cpd.token;
+
+import static org.antlr.v4.runtime.Token.EOF;
+
+import net.sourceforge.pmd.lang.antlr.AntlrTokenManager;
+
+/**
+ * A generic filter for Antlr-based token managers that allows to use comments
+ * to enable / disable analysis of parts of the stream
+ */
+public class AntlrTokenFilter extends BaseTokenFilter<AntlrToken> {
+
+    private boolean discardingHiddenTokens = false;
+
+    /**
+     * Creates a new AntlrTokenFilter
+     * @param tokenManager The token manager from which to retrieve tokens to be filtered
+     */
+    public AntlrTokenFilter(final AntlrTokenManager tokenManager) {
+        super(tokenManager);
+    }
+
+    @Override
+    protected boolean shouldStopProcessing(final AntlrToken currentToken) {
+        return currentToken.getType() == EOF;
+    }
+
+    @Override
+    protected void analyzeToken(final AntlrToken currentToken) {
+        analyzeHiddenTokens(currentToken);
+    }
+
+    @Override
+    protected boolean isLanguageSpecificDiscarding() {
```

Comment:
I'm not sure we should be doing it this way... overriding `isLanguageSpecificDiscarding` for something that is **not** language-specific, but rather Antlr-specific… someone customizing the filter would have to call super on both `analyzeToken` and `isLanguageSpecificDiscarding` to not break the behavior.

I think we should have Antlr behave a lot more like JavaCC, in that the `AntlrTokenManager.getNextToken()` should never return a hidden / special token directly, but always wrapped in the next non-hidden token.